### ISSUE TITLE
when building js_of_ocaml program: "while linking cmo, the external function [...] is not available"

### ISSUE THREAD

**Issue by TyOverby** [ocaml/dune#4027]
## Expected Behavior

```lisp
(executable
 (name foo)
 (modes js)
 (libraries bonsai.web core_kernel)
 (js_of_ocaml (flags --no-sourcemap) (javascript_files example.js))
 (preprocess (pps ppx_jane js_of_ocaml-ppx)))
```

In this executable, which is for , I link "example.js" which defines some external functions.  In the ocaml code, I use 

```ocaml
external my_extern_fun: (float Js.js_array Js.t -> float) -> unit = "my_extern_fun"
```

I expect this program to build.

## Actual Behavior

The build fails with the following 
```
$ dune build @default
File "_none_", line 1: 
Error: Error while linking ui/.pixelmapper.eobjs/byte/dune__exe__Pixelmapper.cmo:
       The external function `my_extern_fun' is not available
```

If I ask for just the .bc.js file, there is no error

```
$ dune build ui/pixel_mapper.bc.js
Done: 0/0 (jobs: 0)
```
and the file is built correctly.

The fact that @default is doing something _extra_ that fails is quite confusing.

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.1
- Version of `ocaml` (output of `ocamlc --version`): 4.10.1

**Comment by nojb** [comment]
This is because the `@default` target is also building the **bytecode** executable, and for bytecode the primitive `my_extern_fun` indeed does not exist. The usual way to handle this is either to define a dummy C primitive for bytecode or to add the bytecode flag `-no-check-prims` to avoid the primitive check.

**Comment by TyOverby** [comment]
> This is because the @default target is also building the bytecode executable

Is this intended?  I explicitly put `(modes js)` in my dune file, so I don't expect anything else to be built when running the default target.  

> The usual way to handle this is either to define a dummy C primitive for bytecode

I don't think that this works either; If I add a `foreign_stubs` field to the dune file, I get this error:

```
Error: Pure bytecode executables cannot contain foreign stubs.
Hint: If you only need to build a native executable use "(modes exe)".
```

> or to add the bytecode flag -no-check-prims to avoid the primitive check.

How is this done?

**Comment by nojb** [comment]
> Is this intended? I explicitly put `(modes js)` in my dune file, so I don't expect anything else to be built when running the default target.

Basically `.bc.js` target requires the `.bc` target as a dependnecy, and `@default` will build basically all targets which have been declared, and this includes `.bc` as well. But maybe there is room for improvement on this matter as this is not the first time this issue has been reported...

> How is this done?

By addin `(link_flags -no-check-prims)` to the `executable` stanza.

**Comment by nojb** [comment]
Hello, can we close this issue?

**Comment by TyOverby** [comment]
> But maybe there is room for improvement on this matter as this is not the first time this issue has been reported...

I agree with you here; this should be fixed in dune.  The @default target just not working is quite confusing.

**Comment by jeremiedimino** [comment]
What about we pass `-no-check-prims` when linking the `.bc` when `modes` doesn't explicitly list `byte`?

**Comment by nojb** [comment]
> What about we pass `-no-check-prims` when linking the `.bc` when `modes` doesn't explicitly list `byte`?

Sounds reasonable to me.

**Comment by hhugo** [comment]
fixed in #5049
